File: 23-Graph/java/03-TraversalTechniques.java
import java.util.*;

class Solution {
    public List<Integer> dfsOfGraph(int V, List<List<Integer>> edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) adj.add(new ArrayList<>());
        for (List<Integer> e : edges) {
            int u = e.get(0), v = e.get(1);
            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[V];
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(0);
        visited[0] = true;

        while (!stack.isEmpty()) {
            int node = stack.pop();
            result.add(node);
            List<Integer> neighbors = adj.get(node);
            for (int i = neighbors.size() - 1; i >= 0; i--) {
                int next = neighbors.get(i);
                if (!visited[next]) {
                    visited[next] = true;
                    stack.push(next);
                }
            }
        }

        return result;
    }

    public List<Integer> bfsOfGraph(int V, List<List<Integer>> edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) adj.add(new ArrayList<>());
        for (List<Integer> e : edges) {
            int u = e.get(0), v = e.get(1);
            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[V];
        Queue<Integer> queue = new LinkedList<>();
        queue.add(0);
        visited[0] = true;

        while (!queue.isEmpty()) {
            int node = queue.poll();
            result.add(node);
            for (int next : adj.get(node)) {
                if (!visited[next]) {
                    visited[next] = true;
                    queue.add(next);
                }
            }
        }

        return result;
    }
}

File: 23-Graph/python/03_traversal_techniques.py
from collections import deque
from typing import List

class Solution:
    def dfsOfGraph(self, V: int, edges: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(V)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        result = []
        visited = [False] * V
        stack = [0]
        visited[0] = True

        while stack:
            node = stack.pop()
            result.append(node)
            for i in range(len(adj[node]) - 1, -1, -1):
                nxt = adj[node][i]
                if not visited[nxt]:
                    visited[nxt] = True
                    stack.append(nxt)

        return result

    def bfsOfGraph(self, V: int, edges: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(V)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        result = []
        visited = [False] * V
        queue = deque([0])
        visited[0] = True

        while queue:
            node = queue.popleft()
            result.append(node)
            for nxt in adj[node]:
                if not visited[nxt]:
                    visited[nxt] = True
                    queue.append(nxt)

        return result
